t/import_csv_file.t

```diff
@@ -0,0 +1,213 @@
+sub deep_extract_keys_from($$) {
+    my ($item1, $item2) = @_;
+    my $result = $item1;  # default is no transformation
+    if ((ref $item1 eq ref {}) && (ref $item2 eq ref {})) {
+        # only keeps keys in item2
+        $result = {};
+        for my $key (keys %$item2) {
+            $result->{$key} = deep_extract_keys_from($item1->{$key}, $item2->{$key});
+        }
+    }
+    if ((ref $item1 eq ref []) && (ref $item2 eq ref []) && (scalar @$item1 == scalar @$item2)) {
+        # note that the case where $item2 is a bag/set is not (yet?) supported
+        $result = [];
+        my $iterator = each_array(@$item1, @$item2);
+        while ( my ($value1, $value2) = $iterator->() ) {
+            push @$result, deep_extract_keys_from($value1, $value2);
+        }
+    }
+    # Note: if $item1 and $item2 does not have same type, 
+    # leave item1 as is, deep comparison will fail
+    return $result
+}
+
+
+# Testing import of a csv file
+{
+    my $import_module = new Test::MockModule('ProductOpener::Import');
+    # mock download image to fetch image in inputs_dir
+    $import_module->mock('download_image', \&fake_download_image);
+    # inputs
+    my $my_excel = $inputs_dir . "test.xlsx";
+    my $columns_fields_json = $inputs_dir . "test.columns_fields.json";
+    my $expected_products_path = $expected_dir . "test_products.perl";
+    # clean data
+    remove_all_products();
+    # step1: parse xls
+    my ($out, $err, $csv_result) = capture (sub {
+        return scalar load_csv_or_excel_file($my_excel);
+    });
+    ok( !$csv_result->{error} );
+    # step2: get columns match
+    my $default_values_ref = {lc => "en", countries => "en"};
+    # this is the file we need
+    my $columns_fields_file = $outputs_dir . "test.columns_fields.sto";
+    create_sto_from_json($columns_fields_json, $columns_fields_file);
+    # step3 convert file
+    my $converted_file = $outputs_dir . "test.converted.csv";
+    my $conv_result;
+    ($out, $err, $conv_result) = capture (sub {
+        return scalar convert_file(
+            $default_values_ref, $my_excel, $columns_fields_file, $converted_file
+        );
+    });
+    ok( !$conv_result->{error} );
+    # step4 import file
+    my $datestring = localtime();
+    my $args = {
+        "user_id" => "test-user",
+        "org_id" => "test-org",
+        "owner_id" => "org-test-org",
+        "csv_file" => $converted_file,
+        "exported_t" => $datestring,
+    };
+    # run
+    ($out, $err) = capture (sub {
+        ProductOpener::Import::import_csv_file($args);
+    });
+    # get all products in db, sorted by code for perdictability
+    my $cursor = execute_query(sub {
+		return get_products_collection()->query({})->sort({code =>1});
+	});
+    my @products = ();
+    while ( my $doc = $cursor->next ) {
+        push(@products, $doc);
+    }
+    # expected values
+    ## no critic (ProhibitStringyEval)
+    my $expected_products = eval(path($expected_products_path)->slurp_raw());
```

Comment:
This works, but I think it would be easier to use the same mechanism that we have for the other tests: we store products as JSON files. So here we would just read all JSON files (one per product) in the directory.

The advantage is that when the expected output change, it's much easier to save the new output as JSON files, and to see diffs. We can also easily see how many products and which products have been affected just by looking at the number of JSON product files that have changed in git.

It's important to note that the expected results will change a lot over time (even if the input CSV does not change) because we keep adding new features like the Nutri-Score, Eco-Score, ingredients parsing etc. that change the resulting product data once it is imported.

So we will need a way to very easily update tests and see diffs.

For instance when I change something in the ingredients parsing, I do this:

perl ingredients.t results expected_tests_results/ingredients
-> this saves the new output product as JSON files over the expected outputs
-> then I can just do "git diff t" and I will see the corresponding changes, to check if they are expected or not.

Reply:
Initially I go for perl because, I wanted to be able to use ignore / bag / supersetof etc from Test::Deep (I didn't know well what to expect).

I think your methodology is also fine, so I can change and go for JSON.
I will however have to clean some fields before comparison, but that can be done in the test part.